orchestra/proc-macro/src/impl_subsystem_ctx_sender.rs

```diff
@@ -494,6 +494,27 @@ pub(crate) fn impl_subsystem_context_trait_for(
 				}
 			}
 
+			async fn recv_signal(&mut self) -> ::std::result::Result<#signal, #error_ty> {
+				self.signals.next().await.ok_or(#support_crate ::OrchestraError::Context(
+					"Signal channel is terminated and empty.".to_owned(),
+				).into())
+			}
+
+			async fn recv_msg(&mut self) -> ::std::result::Result<Self::Message, #error_ty> {
+				loop {
+					if let Some((needs_signals_received, msg)) = self.pending_incoming.take() {
+						self.signals_received.wait_until(|v| v >= needs_signals_received).await;
+						return Ok(msg);
+					}
+					let msg = self.messages.next().await.ok_or(
+						#support_crate ::OrchestraError::Context(
+							"Message channel is terminated and empty.".to_owned()
+						)
+					)?;
+					self.pending_incoming = Some((msg.signals_received, msg.message));
+				}
+			}
```

Comment:
This won't work, since it requires two borrows of self if both streams are to be polled.

Provide a method that produces two adapters.

```
fn incoming_spliter(&mut) -> (SplitOfMessageStream<Self::Message>, SplitOfSignalStream) {
}
```

where both `SplitOfMessageStream` and `SplitOfSignalStream` does `impl Stream` or exposes `async fn recv()`

Reply:
> This won't work, since it requires two borrows of self if both streams are to be polled.

That is exactly what was discussed earlier in this thread :)
[This comment](https://github.com/paritytech/orchestra/pull/66#issuecomment-1825537447) by @sandreim explains how it is supposed to work (and before he explained that I was doubting if it could work myself).